test/plugin/scenarios/spring-jms-5.x-scenario/config/expectedData.yaml

```
@@ -0,0 +1,90 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+segmentItems:
+  - serviceName: spring-jms-5.x-scenario
+    segmentSize: ge 2
+    segments:
+      - segmentId: not null
+        spans:
+          - operationName: ActiveMQ/Queue/test/Producer
+            parentSpanId: 0
+            spanId: 1
+            spanLayer: MQ
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 45
+            isError: false
+            spanType: Exit
+            peer: not blank
+            tags:
+              - { key: mq.broker, value: not blank }
+              - { key: mq.queue, value: test }
+            skipAnalysis: 'false'
+          - operationName: GET:/spring-jms-5.x-scenario/case/spring-jms-scenario
+            parentSpanId: -1
+            spanId: 0
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 1
+            isError: false
+            spanType: Entry
+            peer: ''
+            tags:
+              - { key: url, value: 'http://localhost:8080/spring-jms-5.x-scenario/case/spring-jms-scenario' }
+              - { key: http.method, value: GET }
+              - { key: http.status_code, value: '200' }
+            skipAnalysis: 'false'
+      - segmentId: not null
+        spans:
+          - operationName: ActiveMQ/Queue/test/Consumer
+            parentSpanId: -1
+            spanId: 0
+            spanLayer: MQ
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 46
+            isError: false
+            spanType: Entry
+            peer: not blank
+            tags:
+              - { key: mq.broker, value: not blank }
+              - { key: mq.queue, value: test }
+              - { key: transmission.latency, value: ge 0 }
+            refs:
+              - { parentEndpoint: GET:/spring-jms-5.x-scenario/case/spring-jms-scenario, networkAddress: not null,
+                  refType: CrossProcess, parentSpanId: 1, parentTraceSegmentId: not null,
+                  parentServiceInstance: not null, parentService: not null, traceId: not null }
+            skipAnalysis: 'false'
+      - segmentId: not null
+        spans:
+          - operationName: Jms/queue://test/Execute
```

```suggestion
          - operationName: Jms/queue://test/Execute
```

I can see two entry spans, this and `ActiveMQ/Queue/test/Consumer`. Isn't only one consumer?

---

They are parallel, and ativeMQ does not trace the subsequent actions of consumer, because  JMS use other thread pool to execute listener task.

---

Is that always MQ client dealing with the process? We should consider JMS as a local span if MQ client is the one fetching the message.

---

OK, i will try.

---

Is there any update? Is this practical?

---

yes, but it's not perfect. If we want to trace after MQ consumers span, we need upstream support, such as the ActiveMQ plugin, put the context snapshot somewhere before ending the enhancement method.

---

Do you want jms works without other support? Then what mq is actually working in that scenario?

---

That's not what I mean, but it's OK. At least we can trace what consumers have done.